src/App.tsx

```diff
@@ -1,24 +1,373 @@
+  }, [todos]);
+
+  useEffect(() => {
+    filterTodos();
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [filterParam]);
+
+  useEffect(() => {
+    todoService.getTodos(USER_ID).then(setTodos);
+  }, []);
+
+  function loadTodos() {
+    todoService.getTodos(USER_ID).then((data) => {
+      setTodos(data);
+    }).catch(() => setErrorMsg('Unable to load todos'));
+  }
+
+  function deleteTodo(todoId: number) {
+    setLoadingList((curList) => [...curList, todoId]);
+
+    todoService.removeTodo(todoId)
+      .catch(() => setErrorMsg('Unable to delete a todo'))
+      .then(() => loadTodos())
+      .finally(() => setLoadingList(curList => curList.filter(todo => todo !== todoId)));
+  }
+
+  function addTodo({ title, completed, userId }: Todo) {
+    setDisInput(true);
+
+    todoService.postTodo({ title, completed, userId })
+      .catch(() => setErrorMsg('Unable to add a todo'))
+      .then(newPost => {
+        setTodos(currTodos => [...currTodos, newPost] as Todo[]);
+      }).finally(() => setDisInput(false));
+    setTempTodo(null);
+  }
+
+  function updateTodo(updatedPost: Todo) {
+    setLoadingList((curList) => [...curList, updatedPost.id]);
+    todoService.updateTodo(updatedPost)
+      .catch(() => setErrorMsg('Unable to update a todo'))
+      .then(todo => {
+        setTodos(currTodos => {
+          const newTodo = [...currTodos];
+          const index = newTodo.findIndex(todoe => todoe.id === updatedPost.id);
+
+          newTodo.splice(index, 1, todo as Todo);
+
+          return newTodo;
+        });
+      })
+      .finally(() => {
+        setLoadingList(curList => curList.filter(todo => todo !== updatedPost.id));
+        setRedactedTodo(null);
+      });
+  }
+
+  function onTodoChange(event: FormEvent, {
+    id, title, completed, userId,
+  }: Todo) {
+    event.preventDefault();
+
+    if (title.length < 1) {
+      deleteTodo(id);
+    } else {
+      updateTodo({
+        id, title, completed, userId,
+      });
+    }
+  }
+
+  function redactTodoHandler(todoId: number, todoTitle: string) {
+    setTodoNewTitle(todoTitle); // Спочатку встановлюємо todoNewTitle
+    setRedactedTodo(todoId); // Потім встановлюємо redactedTodo
+  }
+
+  // Пізніше, використовуючи useEffect, ви фокусуєте на введенні, коли redactedTodo змінюється
+  useEffect(() => {
+    if (redactedTodo !== null && todoInputRef.current) {
+      todoInputRef.current.focus();
+    }
+  }, [redactedTodo]);
+
+  useEffect(loadTodos, []);
+
+  function onDeleteCompleted() {
+    todos.filter(todo => todo.completed === true).map(todo => deleteTodo(todo.id));
+  }
+
+  function toggleAll() {
+    if (todos.some(todo => todo.completed === false)) {
+      todos.map(todo => updateTodo({
+        id: todo.id, title: todo.title, completed: true, userId: todo.userId,
+      }));
+    } else {
+      todos.map(todo => updateTodo({
+        id: todo.id, title: todo.title, completed: false, userId: todo.userId,
+      }));
```

Comment:
think about code refactoring. the only difference here only `completed: true or false` , all everything is the same